<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      (Java) 예외처리 &middot; 공부 기록 블로그
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/public/css/main.css">
  <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/public/apple-touch-icon-precomposed.png">
  <link rel="shortcut icon" href="/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">

  
  <script type="text/javascript" async src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_SVG"> </script>
  <script type="text/x-mathjax-config">
MathJax.Hub.Config({ tex2jax: { inlineMath: [ ['$','$'], ["\\(","\\)"] ], processEscapes: true } });
  </script>
  
</head>


  <body>

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox">

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <div class="sidebar-personal-info">
      <div class="sidebar-personal-info-section">
        <!-- <a href="http://gravatar.com/"> -->
          <a href="http://gravatar.com/"></a>
          <!-- <img src="http://www.gravatar.com/avatar/?s=350" title="View on Gravatar" alt="View on Gravatar" /> -->
          <!-- <img src="https://s.gravatar.com/avatar/fedd0cb9bf32323502b502a383136608?s=80" title="View on Gravatar" alt="View on Gravatar" style="margin:auto;" /> -->
          <img src="https://s.gravatar.com/avatar/fedd0cb9bf32323502b502a383136608?s=80" title="View on Gravatar" alt="View on Gravatar" style="margin:auto;" />
          

          
        </a>
      </div>
      <div class="sidebar-personal-info-section" >
        <p>개발자가 되고픈 사람. <br> 제대로 노력해서 1보를 이루자</p>
      </div>
      
    </div>
  </div>

  <nav class="sidebar-nav">
    
      
      
      

      

      <span class="">
        <a class="sidebar-nav-item " href="/">
          Home
        </a>

        
      </span>

    
      
      
      

      

      <span class="foldable">
        <a class="sidebar-nav-item " href="/blog/">
          Blog
        </a>

        
          
            
            
            
              <a class="sidebar-nav-item sidebar-nav-item-sub " href="/blog/categories/">
                Categories
              </a>
          
        
          
            
            
            
              <a class="sidebar-nav-item sidebar-nav-item-sub " href="/blog/tags/">
                Tags
              </a>
          
        
      </span>

    
      
      
      

      

      <span class="">
        <a class="sidebar-nav-item " href="/about/">
          About
        </a>

        
      </span>

    

    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
    &copy; 2020 nobbaggu. This work is liscensed under <a href="http://creativecommons.org/licenses/by-nc/4.0/">CC BY-NC 4.0</a>.
    </p>
  </div>

  <div class="sidebar-item">
    <p>
    Powered by <a href="http://jekyllrb.com">jekyll</a>
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <!-- <a href="/" title="Home" title="공부 기록 블로그">
              <img class="masthead-logo" src=""/>
            </a> -->
            <small>개발자가 되고 싶은 사람</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="post">
  <h1 class="post-title">(Java) 예외처리</h1>
  <span class="post-date">30 Dec 2018</span>
   | 
  
    <a href="/blog/tags/#autocloseable" class="post-tag">autocloseable</a>
  
    <a href="/blog/tags/#catch" class="post-tag">catch</a>
  
    <a href="/blog/tags/#close" class="post-tag">close</a>
  
    <a href="/blog/tags/#exception" class="post-tag">exception</a>
  
    <a href="/blog/tags/#java" class="post-tag">Java</a>
  
    <a href="/blog/tags/#try" class="post-tag">try</a>
  
    <a href="/blog/tags/#예외" class="post-tag">예외</a>
  
    <a href="/blog/tags/#오류" class="post-tag">오류</a>
  
    <a href="/blog/tags/#자바" class="post-tag">자바</a>
  
    <a href="/blog/tags/#처리" class="post-tag">처리</a>
  
  
  <article>
    <p>프로그램을 사용하다 보면 언제나 오류가 생길 가능성이 있다. 컴파일 이전 코드 작성단계에서 발견되는 에러는 큰 문제가 되지않고 수정하면 된다.</p>

<p>그러나 컴파일과정에서 잡지 못한 에러가 있을 수 있다. 이것은 프로그램 실행 중에 문제를 일으킬 소지가 있으며, 따라서 실행오류(Runtime Error)라고 부른다.</p>

<p> </p>

<h1 id="1-실행오류">1. 실행오류</h1>

<hr />

<p>실행오류는 크게 2가지로 나뉜다.</p>

<h3 id="1-오류error">1) 오류(Error)</h3>

<p>말 그대로 오류이다. 사용 가능한 스택 메모리의 부족, 배열의 잘못된 index 접근 등이 있다. 이것은 하드웨어적인 문제이므로 우리가 어떻게 제어할 방법이 없다.</p>

<p>Throwable 클래스를 상속받는 Error클래스는 시스템 오류를 다루는 클래스이다. 그러나 프로그램에서 제어하지 않는다.</p>

<p> </p>

<h3 id="2-예외exception">2) 예외(Exception)</h3>

<p>예외는 우리가 충분히 제어 가능한 경우이다. 네트워크 연결이 끊어진 경우, 파일을 불러오려는데 파일이 비어있는 경우 등이 있다.</p>

<p>Throwable 클래스를 상속받는 Exception클래스와 그 하위클래스들은 예외처리를 할 때 사용한다.</p>

<p> </p>

<p>#</p>

<p><a href="https://SWnomad.com/%ec%98%88%ec%99%b8%ec%b2%98%eb%a6%ac/%ec%a0%9c%eb%aa%a9-%ec%97%86%ec%9d%8c-195/" rel="attachment wp-att-1656"><img class="aligncenter wp-image-1656" src="/images/2018/12/no-name-40.jpg" alt="" width="496" height="240" srcset="/images/2018/12/no-name-40.jpg 738w, /images/2018/12/no-name-40-300x145.jpg 300w" sizes="(max-width: 496px) 100vw, 496px" /></a></p>

<p>모든 예외처리 클래스의 최상위 클래스는 Exception클래스이다. 위 그림에 표시되지 않은 클래스 이외에도 많은 클래스가 있지만 위 클래스들은 사용 빈도가 높은 클래스들이다.</p>

<p>보통 예외가 발생하면 컴파일러가 알려주어 예외처리를 유도하기는 한다. 이 때에는 try-catch문을 사용하여 예외처리를 하면된다. 그런데 RuntimeException 경우는 예외처리를 하지 않아도 컴파일 과정에서 오류가 나지 않는다. 예를 들면 나누기 식에서 분모에 0이 입력되는 경우인데, 실제로 사용자로부터 입력을 받기 전까지는 0이 입력될 지 알 수 없기 때문이다.</p>

<p> </p>

<p> </p>

<h1 id="2-예외-처리하기">2. 예외 처리하기</h1>

<hr />

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      <span class="kt">int</span><span class="o">[]</span> <span class="n">arr</span> <span class="o">=</span> <span class="o">{</span><span class="mi">1</span><span class="o">,</span><span class="mi">2</span><span class="o">,</span><span class="mi">3</span><span class="o">,</span><span class="mi">4</span><span class="o">,</span><span class="mi">5</span><span class="o">};</span>
      
      <span class="c1">//arr는 index가 0~4까지 인데 5를 대입하면 ArrayIndexOutOfBoundsException 발생</span>
      <span class="k">for</span><span class="o">(</span><span class="kt">int</span> <span class="n">i</span><span class="o">=</span><span class="mi">0</span><span class="o">;</span> <span class="n">i</span><span class="o">&lt;=</span><span class="mi">5</span><span class="o">;</span> <span class="n">i</span><span class="o">++)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">arr</span><span class="o">[</span><span class="n">i</span><span class="o">]);</span>
      <span class="o">}</span>
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>위와 같은 예외는 컴파일 과정에서 잡아내지 않는다. 따라서 이후 프로그램 사용중에 런타임 에러에 의해 프로그램이 비정상 종료될 수도 있으므로 프로그래머가 예외처리를 해주어야한다.</p>

<p>예외처리는 try-catch문으로 한다. 위 코드를 예외처리를 하도록 변경하면 아래와 같이 된다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      <span class="kt">int</span><span class="o">[]</span> <span class="n">arr</span> <span class="o">=</span> <span class="o">{</span><span class="mi">1</span><span class="o">,</span><span class="mi">2</span><span class="o">,</span><span class="mi">3</span><span class="o">,</span><span class="mi">4</span><span class="o">,</span><span class="mi">5</span><span class="o">};</span>
      
      
      <span class="k">try</span> <span class="o">{</span><span class="c1">//예외가 일어날 수 있는 코드</span>
         <span class="k">for</span><span class="o">(</span><span class="kt">int</span> <span class="n">i</span><span class="o">=</span><span class="mi">0</span><span class="o">;</span> <span class="n">i</span><span class="o">&lt;=</span><span class="mi">5</span><span class="o">;</span> <span class="n">i</span><span class="o">++)</span> <span class="o">{</span>
            <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">arr</span><span class="o">[</span><span class="n">i</span><span class="o">]);</span>
         <span class="o">}</span>
      <span class="o">}</span><span class="k">catch</span><span class="o">(</span><span class="nc">ArrayIndexOutOfBoundsException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span> <span class="c1">//예외가 일어났을 경우 실행할 코드. 매개변수로는 처리할 예외의 종류</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">);</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"배열 인덱스 범위 오류 예외처리"</span><span class="o">);</span>
      <span class="o">}</span>
      
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>try문 안에는 예외가 발생할 수 있는 코드부분을 넣는다. catch문에는 예외가 일어났을 경우 실행할 코드를 넣어주면 된다. 그리고 catch문의 매개변수로는 처리할 예외를 넣어주면 된다.</p>

<p>위 예제는 컴파일러가 예외를 체크하지 않는다. 그러나 자바에서는 많은 예외들을 컴파일러가 잡아내주어 이에 대해 예외처리를 해주지 않으면 컴파일리 되지 않는다. 파일 입출력이 이와 같은 경우이다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kn">import</span> <span class="nn">java.io.FileInputStream</span><span class="o">;</span> <span class="c1">//파일을 읽고 쓰기위한 스트림</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      <span class="nc">FileInputStream</span> <span class="n">fis</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">FileInputStream</span><span class="o">(</span><span class="s">"myfile.txt"</span><span class="o">);</span> <span class="c1">// FileNotFoundException 예외 발생</span>
      
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>이 예제에서는 읽으려는 파일이 없는경우에 해당하는 FileNotFoundException 예외가 발생하였다. 이 경우도 이 전의 예제와 같이 try-catch문으로 예외처리를 해주면 된다. 빨간줄이 나타나는 부분에 마우스를 가져다 대면 1) 예외를 던지거나 2) try-catch문으로 감싸거나 둘 중 하나를 선택할 수 있다. try-catch문으로 감싸는 것을 선택하면 다음과 같이 코드가 수정된다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kn">import</span> <span class="nn">java.io.FileInputStream</span><span class="o">;</span>
<span class="kn">import</span> <span class="nn">java.io.FileNotFoundException</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      
      <span class="k">try</span> <span class="o">{</span>
         <span class="nc">FileInputStream</span> <span class="n">fis</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">FileInputStream</span><span class="o">(</span><span class="s">"myfile.txt"</span><span class="o">);</span>
      <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="nc">FileNotFoundException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">);</span> <span class="c1">//e.toString() 메소드 호출 -&gt; 예외 메세지 출력</span>
      <span class="o">}</span>

      <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"Program is still running"</span><span class="o">);</span>
      
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>실행결과</p>

<p>java.io.FileNotFoundException: myfile.txt (지정된 파일을 찾을 수 없습니다)</p>

<p>Program is still running</p>

<p>예외 자료형의 참조변수 e를 출력한다는 것은 e.toString()을 출력하는 것이다. 그리고 이 메소드는 해당하는 예외 메세지를 출력하도록 재정의 되어있다.</p>

<p>아무튼 try-catch문으로 예외를 처리해주면 런타임 오류시에도 프로그램의 비정상적인 종료가 일어나지 않는다. 뒷부분에 이를 확인하기 위해 넣어둔 “Program is still running” 구문이 출력되는 것으로 확인할 수 있다.</p>

<p> </p>

<p> </p>

<h1 id="3-finally문">3. finally문</h1>

<hr />

<p>위에서 파일 입출력을 위하여 FileInputStream을 생성하였다. 스트림을 생성한다는 것은 메모리 자원을 사용하는 것이다. 파일 입출력이 끝났다면 이 스트림을 해제하여 해당 자원을 다시 회수하는 것이 올바르다. 위에서는 fis라는 참조변수에 스트림 객체를 저장하였으므로 fis.close(); 메소드 호출을 통하여 스트림을 해제할 수 있다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kn">import</span> <span class="nn">java.io.FileInputStream</span><span class="o">;</span>
<span class="kn">import</span> <span class="nn">java.io.FileNotFoundException</span><span class="o">;</span>
<span class="kn">import</span> <span class="nn">java.io.IOException</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      
      <span class="k">try</span> <span class="o">{</span>
         <span class="nc">FileInputStream</span> <span class="n">fis</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">FileInputStream</span><span class="o">(</span><span class="s">"myfile.txt"</span><span class="o">);</span>
         
         <span class="c1">///////////////////////////</span>
         <span class="k">if</span><span class="o">(</span><span class="n">fis</span> <span class="o">!=</span> <span class="kc">null</span><span class="o">)</span> <span class="o">{</span>
            <span class="k">try</span> <span class="o">{</span>
               <span class="n">fis</span><span class="o">.</span><span class="na">close</span><span class="o">();</span>
            <span class="o">}</span><span class="k">catch</span><span class="o">(</span><span class="nc">IOException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>    <span class="c1">//스트림 종료</span>
               <span class="n">e</span><span class="o">.</span><span class="na">printStackTrace</span><span class="o">();</span>
            <span class="o">}</span>
         <span class="o">}</span>
         <span class="c1">///////////////////////////</span>
         
      <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="nc">FileNotFoundException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">);</span>
      <span class="o">}</span>
      
      <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"Program is still running"</span><span class="o">);</span>
      
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>위 코드는 스트림 종료 메소드 호출을 추가한 것이다. 그런데 지금은 try-catch문이 단순하다. 만약 예외가 발생할 수 있는 부분이 10개라면 10개 모두에 대한 try-catch문에서 모두 스트림 종류 메소드를 호출해 주어야하고 이는 코드를 매우 복잡하게 만들것이다.</p>

<p>이럴 때 사용할 수 있는것이 finally 구문이다. <strong>finally구문은 일단 try문이 실행되었다면 어떤 상황에서도 실행이 된다.</strong></p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kn">import</span> <span class="nn">java.io.FileInputStream</span><span class="o">;</span>
<span class="kn">import</span> <span class="nn">java.io.FileNotFoundException</span><span class="o">;</span>
<span class="kn">import</span> <span class="nn">java.io.IOException</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">ArrayException</span> <span class="o">{</span>
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      
      <span class="nc">FileInputStream</span> <span class="n">fis</span> <span class="o">=</span> <span class="kc">null</span><span class="o">;</span>
      
      <span class="k">try</span> <span class="o">{</span>
         <span class="n">fis</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">FileInputStream</span><span class="o">(</span><span class="s">"myfile.txt"</span><span class="o">);</span>
      <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="nc">FileNotFoundException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">);</span>
         <span class="k">return</span><span class="o">;</span>
      <span class="o">}</span> <span class="k">finally</span> <span class="o">{</span>
         <span class="k">if</span><span class="o">(</span><span class="n">fis</span> <span class="o">!=</span> <span class="kc">null</span><span class="o">)</span> <span class="o">{</span>
            <span class="k">try</span> <span class="o">{</span>
               <span class="n">fis</span><span class="o">.</span><span class="na">close</span><span class="o">();</span>
            <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="nc">IOException</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
               <span class="n">e</span><span class="o">.</span><span class="na">printStackTrace</span><span class="o">();</span>
            <span class="o">}</span>
         <span class="o">}</span>
         
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"This finally block is always executed"</span><span class="o">);</span>
      <span class="o">}</span>
      
   <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>실행결과</p>

<p>java.io.FileNotFoundException: myfile.txt (지정된 파일을 찾을 수 없습니다)</p>

<p>This finally block is always executed</p>

<p>예외가 발생하여 catch로 넘어가고 return문을 통해 프록램이 종료되도록 하더라도 finally문이 실행 된 다음에 프로그램이 종료된다.</p>

<p> </p>

<p> </p>

<h1 id="4-autocloseable-인터페이스">4. AutoCloseable 인터페이스</h1>

<hr />

<p>사실 Java7 부터는 FileInputStream이 명시적으로 close() 메소드 호출을 통해 닫아주지 않더라도 자동으로 닫히도록 구현되어있다.</p>

<p>AutoCloseable 인터페이스는 close() 추상메소드를 가지고 있고, 이를 구현한 클래스들은 close()메소드를 재정의 하여야한다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">AutoCloseTest</span> <span class="kd">implements</span> <span class="nc">AutoCloseable</span> <span class="o">{</span>

   <span class="nd">@Override</span>
   <span class="kd">public</span> <span class="kt">void</span> <span class="nf">close</span><span class="o">()</span> <span class="kd">throws</span> <span class="nc">Exception</span> <span class="o">{</span>
      <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"close() is called automatically"</span><span class="o">);</span> <span class="c1">//호출되면 해당 문장 출력    </span>
   <span class="o">}</span>
   
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      <span class="k">try</span><span class="o">(</span><span class="nc">AutoCloseTest</span> <span class="n">obj</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">AutoCloseTest</span><span class="o">()){</span>
         <span class="c1">//No Code</span>
      <span class="o">}</span><span class="k">catch</span><span class="o">(</span><span class="nc">Exception</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">);</span>
      <span class="o">}</span>
   <span class="o">}</span>

<span class="o">}</span>


</code></pre></div></div>

<p>실행결과</p>

<p>close() is called automatically</p>

<p>AutoCloseTest 클래스는 AutoCloseable 인터페이스를 구현하였고 close() 메소드를 재정의하였다. 이후 try문에서 AutoCloseTest로 객체를 하나 만들어 넘겨주었고 close() 메소드가 자동으로 호출되는 것을 확인할 수 있다.</p>

<p>catch문에서도 close 메소드가 호출되는지 확인해야한다. 위 코드에서 //No Code 부분에 throw new Exception(); 이라고 치면 강제로 예외를 만들어내어 catch문을 실행시킬 수 있다.</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">package</span> <span class="nn">exception</span><span class="o">;</span>

<span class="kd">public</span> <span class="kd">class</span> <span class="nc">AutoCloseTest</span> <span class="kd">implements</span> <span class="nc">AutoCloseable</span> <span class="o">{</span>

   <span class="nd">@Override</span>
   <span class="kd">public</span> <span class="kt">void</span> <span class="nf">close</span><span class="o">()</span> <span class="kd">throws</span> <span class="nc">Exception</span> <span class="o">{</span>
      <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"close() is called automatically"</span><span class="o">);</span> <span class="c1">//호출되면 해당 문장 출력    </span>
   <span class="o">}</span>
   
   <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span><span class="o">(</span><span class="nc">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
      <span class="k">try</span><span class="o">(</span><span class="nc">AutoCloseTest</span> <span class="n">obj</span> <span class="o">=</span> <span class="k">new</span> <span class="nc">AutoCloseTest</span><span class="o">()){</span>
         <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">();</span> <span class="c1">//강제로 예외 발생시켜 catch문 실행</span>
      <span class="o">}</span><span class="k">catch</span><span class="o">(</span><span class="nc">Exception</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
         <span class="nc">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"This is exception"</span><span class="o">);</span>
      <span class="o">}</span>
   <span class="o">}</span>

<span class="o">}</span>
</code></pre></div></div>

<p>실행결과</p>

<p>close() is called automatically</p>

<p>This is exception</p>


  </article>
</div>

<div class="related">
  <h2>Related Posts</h2>
  <ul class="related-posts">
    
      <li>
        <h3>
          <a href="/2020/03/09/%ED%95%A8%EC%88%98/">
            (클린코드) Chapter03 - 함수
            <small>09 Mar 2020</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/2020/03/08/%EC%9D%98%EB%AF%B8-%EC%9E%88%EB%8A%94-%EC%9D%B4%EB%A6%84/">
            (클린코드) Chapter02 - 의미 있는 이름
            <small>08 Mar 2020</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/2020/03/07/%EA%B9%A8%EB%81%97%ED%95%9C%EC%BD%94%EB%93%9C/">
            (클린코드) Chapter01 - 깨끗한 코드
            <small>07 Mar 2020</small>
          </a>
        </h3>
      </li>
    
  </ul>
</div>


<div class="comments">
  <h2>Comments</h2>
  <div id="disqus_thread"></div>
  <script>
var disqus_config = function () {
  this.page.url = 'http://localhost:4000/2018/12/30/exception/'; // Replace PAGE_URL with your page's canonical URL variable
  this.page.identifier = '/2018/12/30/exception'; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
(function() {
  var d = document, s = d.createElement('script');
  s.src = '//wkimdev-gitblog.disqus.com/embed.js';
  s.setAttribute('data-timestamp', +new Date());
  (d.head || d.body).appendChild(s);
})();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>
</div>


      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if (target === toggle) {
            checkbox.checked = !checkbox.checked;
            e.preventDefault();
          } else if (checkbox.checked && !sidebar.contains(target)) {
            /* click outside the sidebar when sidebar is open */
            checkbox.checked = false;
          }
        }, false);
      })(document);
    </script>
    
    <script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){ (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m) })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-00000000-1', 'auto');
ga('send', 'pageview');
    </script>
    
  </body>
  
  <script id="dsq-count-scr" src="//wkimdev-gitblog.disqus.com/count.js" async></script>
  
</html>
